### PR TITLE
added pyright and fixed a few issues

### PR DESCRIPTION
**Pull request by tarekziade**
`pyright` is Microsoft's static type checker (MIT)

It detected two major bugs in our codebase (yay):

- a serializing error in the Mongo connector
- a coroutine that was never awaited in `byoc.py`
 
This patch fixes them and ignore a few warnings. It's now plugged into `make lint`

### CODE REVIEW

File: connectors/sources/s3.py [hunk 1]
```diff
@@ -183,7 +185,7 @@ async def get_docs(self, filtering=None):
             dictionary: Document from Amazon S3.
         """
         bucket_list = self.buckets if self.buckets != ["*"] else self.get_bucket_list()
-        page_size = int(self.configuration.get("page_size", DEFAULT_PAGE_SIZE))
+        page_size = self.configuration["page_size"]
```
Comment (wangch079) [inline connectors/sources/s3.py:188]: Not quite sure about this change, it seems it remove the usage of `DEFAULT_PAGE_SIZE`

File: connectors/sources/mysql.py [hunk 2]
```diff
@@ -243,8 +243,7 @@ async def _connect(self, query, fetch_many=False, **query_kwargs):
 
         query_kwargs["database"] = self.database
         formatted_query = query.format(**query_kwargs)
-        size = int(self.configuration.get("fetch_size", DEFAULT_FETCH_SIZE))
-
+        size = self.configuration["fetch_size"]
```
Comment (wangch079) [inline connectors/sources/mysql.py:246]: It removes `DEFAULT_FETCH_SIZE`. What if there's no key `fetch_size` in `self.configuration`

File: connectors/sources/generic_database.py [hunk 3]
```diff
@@ -152,6 +152,9 @@ async def execute_query(self, query_name, fetch_many=False, **query_kwargs):
         Yields:
             list: Column names and query response
         """
+        if self.queries is None:
+            raise NotImplementedError()
```
Comment (wangch079) [inline connectors/sources/generic_database.py:156]: It makes sense to raise error if `self.queries` is None, but `NotImplementedError`?

File: connectors/byoc.py [hunk 4]
```diff
@@ -693,7 +693,7 @@ async def sync(self, sync_job_index, source_klass, elastic_server, bulk_options)
                 logger.debug(
                     f"No change in {self.service_type} data provider, skipping..."
                 )
-                self._sync_done(job, JobStatus.COMPLETED, {}, start_time)
+                await self._sync_done(job, JobStatus.COMPLETED, {}, start_time)
```
Comment (wangch079) [inline connectors/byoc.py:696]: Good catch

**Review by wangch079 — COMMENTED** [review]
> - a coroutine that was never awaited in byoc.py

This is extremely useful.

File: connectors/sources/mongo.py [hunk 5]
```diff
@@ -73,7 +73,7 @@ def serialize(self, doc):
         def _serialize(value):
             if isinstance(value, ObjectId):
                 value = str(value)
-            if isinstance(value, (list, tuple)):
+            elif isinstance(value, (list, tuple)):
```
Comment (artem-shelkovnikov) [inline connectors/sources/mongo.py:76]: Oh wow

**Review by artem-shelkovnikov — COMMENTED** [review]
I've seen several "# pyright: ignore" added - what about them? Is the rule okay but bad for these lines? Are we planning to address these later?

File: connectors/sources/s3.py (hunk 1 above)
Comment (tarekziade) [reply connectors/sources/s3.py:188]: no, it's still used but done directly in the base class now on init, this code is now useless.
https://github.com/elastic/connectors-python/blob/main/connectors/source.py#L124

File: connectors/sources/generic_database.py (hunk 3 above)
Comment (tarekziade) [reply connectors/sources/generic_database.py:156]: `self.queries` is set to None in this base class. If the concrete class does not set it and this function is called, it won't work. So in that case it means the concrete class is incomplete. And for incomplete concrete class, you use `NotImplementedError`

**Comment by tarekziade** [comment]
> I've seen several "# pyright: ignore" added - what about them? Is the rule okay but bad for these lines? Are we planning to address these later?

It happens when we use things not as intended. When we derive `asyncio.Queue` we use `_members` that are marked as being private to that class in the typing definitions in the stdlib.  And it's a bit of a grey area: should we rewrite the whole class? or use composition? To address this, we would need to rewrite that class in particular

**Review by artem-shelkovnikov — APPROVED** [review]

**Review by wangch079 — APPROVED** [review]

**Pull request merged**